### PR TITLE
[mono] Add Android app sample using mono runtime

### PR DESCRIPTION
**Pull request by mdh1418**
## Summary

This PR looks to port over the Android [sample](https://github.com/dotnet/runtime/tree/master/src/mono/netcore/sample/Android ) from `dotnet/runtime` to highlight the concept of building applications on top of the mono runtime. 
There are several dependencies such as Android SDK, Android NDK, an Android emulator required to run the sample with `dotnet publish`.

The sample will depend on a NuGet Package containing the files that build the app and embed the mono runtime into the app, and the code that will build that package can be found here https://github.com/dotnet/runtime/pull/43216

Currently, it is rebased off of https://github.com/dotnet/samples/pull/3931 and will be updated after the iOS sample is merged.

### CODE REVIEW

File: core/mono-samples/Android/AndroidSampleApp.cs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Runtime.InteropServices;
+
+// it's not part of the BCL but runtime needs it for native-to-managed callbacks in AOT
+// To be replaced with NativeCallableAttribute
+public class MonoPInvokeCallbackAttribute : Attribute
```
Comment (EgorBo) [inline core/mono-samples/Android/AndroidSampleApp.cs:11]: We probably can get rid of it using Function Pointers actually but I'm not sure, last time I checked - callbacks still needed it.

File: core/mono-samples/Android/AndroidSampleApp.csproj [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    <OutputType>Exe</OutputType>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.NET.Runtime.Android.Sample.Mono" Version="6.0.0-*" GeneratePathProperty="true" />
+  </ItemGroup>
+
+  <Target Name="BuildNativeLib" BeforeTargets="Publish">
+    <Exec Command="cmake -DCMAKE_TOOLCHAIN_FILE=$ANDROID_NDK_ROOT/build/cmake/android.toolchain.cmake -DANDROID_ABI=x86_64 -DANDROID_STL=none -DANDROID_NATIVE_API_LEVEL=21 -B monodroid -DCMAKE_BUILD_TYPE=MinSizeRel -DCMAKE_C_FLAGS=%22-s -Wno-unused-command-line-argument%22" />
+    <Exec Command="cmake --build monodroid --config MinSizeRel" />
```
Comment (EgorBo) [inline core/mono-samples/Android/AndroidSampleApp.csproj:14]: `DANDROID_ABI=x86_64` limits it for x64 only

File: core/mono-samples/Android/AndroidSampleApp.cs (hunk 1 above)
Comment (EgorBo) [reply core/mono-samples/Android/AndroidSampleApp.cs:11]: And it's not needed for JIT (which is the only option for Android atm)

File: core/mono-samples/Android/native-lib.c [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+androidRegisterNameGreet (void* ptr)
+{
+    greetHandlerPtr = ptr;
+}
+
+int myAge() {
+    JNIEnv* jniEnv = GetJniEnv();
+
+    jclass mainActivity = (*jniEnv)->FindClass (jniEnv, "net/dot/MainActivity");
+    jmethodID getAge = (*jniEnv)->GetStaticMethodID(jniEnv, mainActivity, "getAge", "()I");
```
Comment (EgorBo) [inline core/mono-samples/Android/native-lib.c:64]: NOTE: FindClass won't work for Activity in non-UI thread afair.

File: core/mono-samples/Android/AndroidSampleApp.cs [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+public static class AndroidSampleApp
+{
+    [DllImport("native-lib")]
+    private extern static int myAge();
+
+    [DllImport("native-lib")]
+    private extern static void androidRegisterCounterIncrement(Func<string> del);
+    private static Func<string> incrementHandler = null;
```
Comment (EgorBo) [inline core/mono-samples/Android/AndroidSampleApp.cs:23]: You can get rid of `incrementHandler` variable - we don't need to save static callback into it. (same for `nameHandler`)

File: core/mono-samples/Android/AndroidSampleApp.cs [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    private extern static void androidRegisterNameGreet(Func<string,string> del);
+    private static Func<string,string> nameHandler = null;
+
+    private static int counter = 1;
+
+    // Called by native code, see native-lib.c
+    [MonoPInvokeCallback(typeof (Func<string>))]
+    private static string IncrementCounter()
+    {
+        return "Clicked " + counter++ + " times!";
```
Comment (EgorBo) [inline core/mono-samples/Android/AndroidSampleApp.cs:35]: `$"Clicked {counter++} times!"`

File: core/mono-samples/Android/AndroidSampleApp.csproj [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+    <AndroidAppBuilderTask
+      RuntimeIdentifier="android-x64"
+      SourceDir="$(OutputPath)"
+      MonoRuntimeHeaders="%(AndroidRuntimePack.PackageDirectory)\runtimes\android-x64\native\include\mono-2.0"
+      MainLibraryFileName="$(AssemblyName).dll"
+      ProjectName="HelloAndroid"
+      OutputDir="$(ApkDir)"
+      StripDebugSymbols="True"
+      KeyStorePath="/Users/mdhwang/.android/"
```
Comment (EgorBo) [inline core/mono-samples/Android/AndroidSampleApp.csproj:50]: TEMP or `~`(HOME) directory?

File: core/mono-samples/Android/AndroidSampleApp.csproj [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+      NativeMainSource="$(MSBuildThisFileDirectory)MainActivity.java">
+      <Output TaskParameter="ApkBundlePath" PropertyName="ApkBundlePath" />
+      <Output TaskParameter="ApkPackageId" PropertyName="ApkPackageId" />
+    </AndroidAppBuilderTask>
+
+    <Exec Command="$(AdbTool) kill-server"/>
+    <Exec Command="$(AdbTool) start-server"/>
+    <Exec Command="$(AdbTool) logcat -c" ContinueOnError="WarnAndContinue" />
+    <Exec Command="$(AdbTool) install $(ApkDir)/bin/HelloAndroid.apk" />
+    <Exec Condition="'$(RunActivity)' != 'true'" Command="$(AdbTool) shell am instrument -w net.dot.HelloAndroid/net.dot.MonoRunner" />
```
Comment (EgorBo) [inline core/mono-samples/Android/AndroidSampleApp.csproj:60]: Remove this line - it's for headless launch, I guess we don't need it for samples. (and remove $(RunActivity))

File: core/mono-samples/Android/MainActivity.java [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+import android.graphics.Color;
+
+public class MainActivity extends Activity
+{
+    static
+    {
+        System.loadLibrary("native-lib");
+    }
+
+    public static int getAge()
```
Comment (EgorBo) [inline core/mono-samples/Android/MainActivity.java:24]: NIT: add a comment: called from JNI

File: core/mono-samples/Android/AndroidSampleApp.csproj (hunk 6 above)
Comment (mdh1418) [reply core/mono-samples/Android/AndroidSampleApp.csproj:50]: Thanks, didn't know that `$(HOME)` worked. `~/.android` didn't evaluate in the ApkBuilder.

File: core/mono-samples/Android/AndroidSampleApp.csproj (hunk 7 above)
Comment (mdh1418) [reply core/mono-samples/Android/AndroidSampleApp.csproj]: Thanks!

File: core/mono-samples/Android/AndroidSampleApp.cs (hunk 5 above)
Comment (mdh1418) [reply core/mono-samples/Android/AndroidSampleApp.cs]: Done

File: core/mono-samples/Android/AndroidSampleApp.cs (hunk 4 above)
Comment (mdh1418) [reply core/mono-samples/Android/AndroidSampleApp.cs]: Thanks

File: core/mono-samples/Android/AndroidSampleApp.cs (hunk 1 above)
Comment (mdh1418) [reply core/mono-samples/Android/AndroidSampleApp.cs]: It works after removing this block as well as the attributes. Does this still qualify as showing how the embedded mono runtime allows for C# and C to call each other with just the function pointers and `Dllimport`?

File: core/mono-samples/Android/AndroidSampleApp.csproj (hunk 2 above)
Comment (mdh1418) [reply core/mono-samples/Android/AndroidSampleApp.csproj:21]: Soft coded the architecture to allow for other architectures to be targeted.

**Comment by mdh1418** [comment]
@adegeo Can the check run `dotnet publish` for these mono-samples rather than `dotnet build` https://github.com/dotnet/samples/blob/2baf6a963625a043457e187a3def8b68356cad34/.github/workflows/dependencies/Get-MSBuildResults.ps1#L138?

**Review by steveisok — APPROVED** [review]

**Comment by adegeo** [comment]
@mdh1418 

> @adegeo Andy De George FTE Can the check run `dotnet publish` for these mono-samples rather than `dotnet build`
> 
> https://github.com/dotnet/samples/blob/2baf6a963625a043457e187a3def8b68356cad34/.github/workflows/dependencies/Get-MSBuildResults.ps1#L138
> 
> ?

I just updated the script to allow custom commands. Add the `snippets.5000.json` file to the same folder as the `AndroidSampleApp.csproj` file with this content:

```json
{
    "host": "custom",
    "command": "dotnet publish \"$projectFile\""
}
```

I'm not sure if you need to pull in the latest from master or not.

**Comment by mdh1418** [comment]
Thanks @adegeo! I think the problems for these mono samples was the github actions runner being `windows-latest`. I don't think the support to build mono on Windows is there yet, so I opted to add a MacOS lane to verify that the samples build properly via `dotnet publish`. The related change https://github.com/dotnet/samples/pull/3953 would only run for PRs with changes to the mono sample.

**Pull request merged**